### ISSUE TITLE
Low output

### ISSUE THREAD

**Issue by Bewildert** [benjjneb/dada2#753]
Dear everyone,

I am a Ph.D. student and new to the whole world of genomics. Nevertheless, there are some NGS methods included in my project so I tried to use this very impressive package.

Now I tried to modify the tutorial script to my needs but it somehow shows far fewer ASVs than I would expect. A friend of mine used his pipeline to calculate a traditional OTU table and for my dataset, he found about 20000 OUTs while I only find about 2500 ASVs. 

I really don`t know what I am doing wrong, so I am very happy about any comments.

This is the script I used:

 `####DADA2 for slow Pcs ;)

### modified by Benjamin Wilden

#R 3.5.0 required
###ORIGINAL PIPELINE DADA2 Tutorial (1.8) at https://benjjneb.github.io/dada2/tutorial.html

#install if necessary
source("https://bioconductor.org/biocLite.R")
biocLite()
source("https://bioconductor.org/biocLite.R")
biocLite("dada2")
#packageVersion("dada2")

library(Rcpp)
library(dada2)

setwd(choose.dir())

# File parsing
pathF <- "R1" # CHANGE ME to the directory containing your demultiplexed forward-read fastqs
pathR <- "R2" # CHANGE ME ...
filtpathF <- file.path(pathF, "filtered") # Filtered forward files go into the pathF/filtered/ subdirectory
filtpathR <- file.path(pathR, "filtered") # ...
fastqFs <- sort(list.files(pathF, pattern="fastq.gz"))#, full.names = TRUE
fastqRs <- sort(list.files(pathR, pattern="fastq.gz"))#, full.names = TRUE

#Inspecting read quality profiles; enable full.names to plot
pdf("Quality Profile.pdf")
plotQualityProfile(fastqFs[1:2] ) 
plotQualityProfile(fastqRs[1:2])
dev.off()

if(length(fastqFs) != length(fastqRs)) stop("Forward and reverse files do not match.")
# Filtering: THESE PARAMETERS ARENT OPTIMAL FOR ALL DATASETS
out<-filterAndTrim(fwd=file.path(pathF, fastqFs), filt=file.path(filtpathF, fastqFs),
              rev=file.path(pathR, fastqRs), filt.rev=file.path(filtpathR, fastqRs),
               maxEE=2, truncQ=11, maxN=0, rm.phix=TRUE,
              compress=TRUE, verbose=TRUE, multithread=TRUE, trimLeft = 18 ) #truncLen=c(280,280)to exclude bad quality parts; ,trimLeft = c(FWD_PRIMER_LEN, REV_PRIMER_LEN); for me 16S=c(17,21) 18S= 18
head(out)

# File parsing
filtpathF <- "R1/filtered" # CHANGE ME to the directory containing your filtered forward fastqs
filtpathR <- "R2/filtered" # CHANGE ME ...
filtFs <- list.files(filtpathF, pattern="fastq.gz", full.names = TRUE)
filtRs <- list.files(filtpathR, pattern="fastq.gz", full.names = TRUE)
sample.names <- sapply(strsplit(basename(filtFs), "_R"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
sample.namesR <- sapply(strsplit(basename(filtRs), "_R"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
if(!identical(sample.names, sample.namesR)) stop("Forward and reverse files do not match.")
names(filtFs) <- sample.names
names(filtRs) <- sample.names
set.seed(100)
# Learn forward error rates
errF <- learnErrors(filtFs, nbases=1e8,verbose=TRUE, multithread=TRUE)
# Learn reverse error rates
errR <- learnErrors(filtRs, nbases=1e8,verbose=TRUE, multithread=TRUE)

#Check errors
pdf("Errorplot.pdf")
plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)
dev.off()

# Sample inference and merger of paired-end reads
mergers <- vector("list", length(sample.names))
names(mergers) <- sample.names
for(sam in sample.names) {
  cat("Processing:", sam, "\n")
    derepF <- derepFastq(filtFs[[sam]])
    ddF <- dada(derepF, err=errF, multithread=TRUE)
    derepR <- derepFastq(filtRs[[sam]])
    ddR <- dada(derepR, err=errR, multithread=TRUE)
    merger <- mergePairs(ddF, derepF, ddR, derepR)
    mergers[[sam]] <- merger
}
rm(derepF); rm(derepR)
head(mergers[[1]])

# Construct sequence table and remove chimeras
seqtab <- makeSequenceTable(mergers)
saveRDS(seqtab, "seqtab.rds") # CHANGE ME to where you want sequence table saved

# Merge multiple runs (if necessary)
st1 <- readRDS("seqtab.rds")
#st2 <- readRDS("path/to/run2/output/seqtab.rds")
#st3 <- readRDS("path/to/run3/output/seqtab.rds")
st.all <- st1 #mergeSequenceTables(st1, st2, st3)
# Remove chimeras
seqtab <- removeBimeraDenovo(st.all, method="consensus", multithread=TRUE)#check out method pool
# Assign taxonomy
tax <- assignTaxonomy(seqtab, "silva_nr_v132_train_set.fa.gz", multithread=TRUE)
tax <- addSpecies(tax, "silva_species_assignment_v132.fa.gz")
taxa.print <- tax # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)

#Tracking reads through the pipeline
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(mergers, getN), rowSums(seqtab))
colnames(track) <- c("input", "filtered",  "merged", "nonchim")
rownames(track) <- sample.names
head(track)

# Write to disk
saveRDS(seqtab, "seqtab_final.rds") # CHANGE ME to where you want sequence table saved
saveRDS(tax, "tax_final.rds") # CHANGE ME ...
`

I tried to play around with the parameters but it seems there is something more fundamentally wrong.

Thanks in advance!

Best wishes,

Benjamin

**Comment by benjjneb** [comment]
What fraction of your reads make it through the pipeline? I.e. what does `head(track)` output?

> A friend of mine used his pipeline to calculate a traditional OTU table and for my dataset, he found about 20000 OUTs while I only find about 2500 ASVs.

Many of the traditional OTU pipelines didn't do a very good job of controlling for PCR/sequencing errors, and could output numbers of OTUs far higher than were present in the data, e.g. https://msystems.asm.org/content/1/1/e00003-15.abstract

Typically DADA2 does a better job of avoiding spurious output, and so can produce a lot fewer ASVs than you might expect based on some older pipelines.

I don't know that's what is going on in your data, but there's a good chance it is. I would recommend that you reconsider your filtering, especially `trimLeft` and `truncLen`, as those are the two parameters that can have the biggest impact on your output.  See this presentation for more info and some exercises on setting those parameters appropriately: http://htmlpreview.github.io/?https://github.com/benjjneb/NCSUWorkshop/blob/master/Lecture3_dada2_workflow.html

**Comment by Bewildert** [comment]
Thank you for the fast response.

I always used trimLeft to eliminate primers. But truncLen further reduced the output drastically. I will post the head(track) for the trimmed version as well but it is still running.

This is the output without truncLen:

`> head(track)
                                   input filtered merged nonchim
MI.M03992_0325.001.FLD0194.16S_13 103928    87688   2727    2727
MI.M03992_0325.001.FLD0195.16S_25 111669    94981   4598    4593
MI.M03992_0325.001.FLD0196.16S_37 114769    96325   5388    5388
MI.M03992_0325.001.FLD0197.16S_49 123934   103523   4416    4414
MI.M03992_0325.001.FLD0199.16S_1  119298   100951  23392   23274
MI.M03992_0325.001.FLD0202.16S_14 126349   106839   3336    3336`

If I take the duplicates into account we see:
> dim(seqtab)
[1]   52 2857

So I guess this would be almost 3000 ASVs. As I said there were about 20000 OTUs found. Isn't that a bit too much?

Thanks again.

**Comment by benjjneb** [comment]
If I'm reading that table right (the formatting is messed up) you are lowing most of your reads at the merger step. That is an indication that something isn't working as expected. Usually this is a consequence of poor overlap between the reads in longer amplicons.

What amplicon are you sequencing? What are the primers? Is this 2x300 sequencing with the primers included?

**Comment by Bewildert** [comment]
Yes, you are right. Let me try another time:

> > head(track)                                         input filtered merged nonchim 
MI.M03992_0325.001.FLD0194.16S_13 103928 87688 2727 2727 MI.M03992_0325.001.FLD0195.16S_25 111669 94981 4598 4593 MI.M03992_0325.001.FLD0196.16S_37 114769 96325 5388 5388 MI.M03992_0325.001.FLD0197.16S_49 123934 103523 4416 4414 MI.M03992_0325.001.FLD0199.16S_1 119298 100951 23392 23274 MI.M03992_0325.001.FLD0202.16S_14 126349 106839 3336 3336

We sequenced 16S with standard 341F-805R 2*300.
But how is that not overlapping? I guess the quality was quite okay:
[Quality Profile 16S.pdf](https://github.com/benjjneb/dada2/files/3161500/Quality.Profile.16S.pdf)

For trimming, I now used c(260,210) and as always trimLeft c(17,21) to get rid of the primers. Now only about 30% remain after the filtering. That is what I saw before and why I did not use the trimming.

**Comment by benjjneb** [comment]
What happens when you increase to `truncLen=c(285, 225)`? I.e. what does the `head(track)` output then?

**Comment by Bewildert** [comment]
Okay, with the truncLen=c(260, 210) I had:

>                                                              input filtered merged nonchim
MI.M03992_0325.001.FLD0194.16S_13 103928    29307   2112    2035
MI.M03992_0325.001.FLD0195.16S_25 111669    32314   1556    1556
MI.M03992_0325.001.FLD0196.16S_37 114769    32634   4454    4280
MI.M03992_0325.001.FLD0197.16S_49 123934    33598   5301    5163
MI.M03992_0325.001.FLD0199.16S_1  119298    34109   2718    2653
MI.M03992_0325.001.FLD0202.16S_14 126349    35372   2738    2642
> dim(seqtab)
[1]   52 7792

and truncLen=c(285, 225) is now:
      					                           input filtered merged nonchim
MI.M03992_0325.001.FLD0194.16S_13 103928    18715   4605    4549
MI.M03992_0325.001.FLD0195.16S_25 111669    20765   6852    6754
MI.M03992_0325.001.FLD0196.16S_37 114769    21334   8722    8584
MI.M03992_0325.001.FLD0197.16S_49 123934    21645   8624    8596
MI.M03992_0325.001.FLD0199.16S_1  119298    22146   6080    5957
MI.M03992_0325.001.FLD0202.16S_14 126349    22326   5693    5654

> dim(seqtab)
[1]    52 14102

  This is obviously better than everything I got so far. Thanks a lot! Somehow I think this should be within the range of results you mentioned (about 25% less than OTUs). What do you think?

**Comment by benjjneb** [comment]
That seems to be an improvement, but I would still be concerned about the low rate (~33%) of successful merging you are getting.

Perhaps one more thing to try would be to run the `dada` step in pseudo-pooling mode, i.e. `dada(..., pool="pseudo")` and see if that increases the merging rate.

**Comment by Bewildert** [comment]
I tried the pool="pseudo" but the results are almost the same. No improvement.

**Comment by benjjneb** [comment]
Hmm. Would you be willing to share one of your samples I could take a closer look at?

**Comment by Bewildert** [comment]
Just one sample? Sure. If you need more let me know. 
[MI.M03992_0325.001.FLD0194.16S_13_R1.fastq.gz](https://github.com/benjjneb/dada2/files/3174449/MI.M03992_0325.001.FLD0194.16S_13_R1.fastq.gz)
[MI.M03992_0325.001.FLD0194.16S_13_R2.fastq.gz](https://github.com/benjjneb/dada2/files/3174454/MI.M03992_0325.001.FLD0194.16S_13_R2.fastq.gz)

Thank you so much for your effort.

**Comment by benjjneb** [comment]
It appears that you have some kind of heterogenity spacer setup, which is causing the problems you are seeing. See the output of the following:

```
fnF <- "MI.M03992_0325.001.FLD0194.16S_13_R1.fastq.gz"
substr(getSequences(derepFastq(fnF))[1:15], 1,25)
```
 >[1] "TCCTACGGGAGGCTGCAGTGGGGAA" "TCCTACGGGGGGCTGCAGTGGGGAA"
 >[3] "TCCTACGGGTGGCTGCAGTGGGGAA" "CCTACGGGAGGCTGCAGTGGGGAAT"
 >[5] "TCCTACGGGAGGCAGCAGTGGGGAA" "TCCTACGGGGGGCAGCAGTGGGGAA"
 >[7] "CCTACGGGAGGCAGCAGTGGGGAAT" "TCCTACGGGGGGCTGCAGTGGGGAA"
 >[9] "ACCCTACGGGGGGCTGCAGTGGGGA" "TCCTACGGGCGGCTGCAGTGGGGAA"
>[11] "CTACCTACGGGGGGCAGCAGTGGGG" "CTACCTACGGGGGGCTGCAGTGGGG"
>[13] "TCCTACGGGGGGCAGCAGTGGGGAA" "TCCTACGGGGGGCTGCAGTAAGGAA"
>[15] "CCTACGGGGGGCAGCAGTGGGGAAT"

Notice how before the 341F primer (`CCTACGGGNGGCWGCAG`) there are appears 0-3 bases, sometimes nothing, sometimes `T`, sometimes `AC`, sometimes `CTA`. Because the location of the primer in the same in every read, the standard `trimLeft` strategy to remove primers is not effective, which downstream manifests itself as a low merge rate. You need to use an external program like cutadapt or trimmomatic to identify where the primers are and remove them before proceeding with the dada2 pipeline.

**Comment by Bewildert** [comment]
This is something I did not expect. Well, okay thank you for showing me.

I need to trim everything with cutadapt but just for the quick view, I trimmed the primers plus 5 bp so the additional bp should be gone. 

>                          input filtered merged nonchim
MI.M03992_0325.001.FLD0194.16S_13 103928    18731   4704    4653
MI.M03992_0325.001.FLD0195.16S_25 111669    20782   6947    6822
MI.M03992_0325.001.FLD0196.16S_37 114769    21362   8847    8744
MI.M03992_0325.001.FLD0197.16S_49 123934    21658   8557    8530
MI.M03992_0325.001.FLD0199.16S_1  119298    22170   6130    6024
MI.M03992_0325.001.FLD0202.16S_14 126349    22347   5754    5646

> dim(seqtab)
[1]    52 14139

So no real improvement. I will see if cutadapt generates a major difference.
